zookeeper-server/src/test/java/org/apache/zookeeper/server/X509AuthFailureTest.java

```diff
@@ -0,0 +1,242 @@
+        clientX509Util = new ClientX509Util();
+        String testDataPath = System.getProperty("test.data.dir", "src/test/resources/data");
+        System.setProperty(ServerCnxnFactory.ZOOKEEPER_SERVER_CNXN_FACTORY, "org.apache.zookeeper.server.NettyServerCnxnFactory");
+        System.setProperty(ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET, "org.apache.zookeeper.ClientCnxnSocketNetty");
+        System.setProperty(ZKClientConfig.SECURE_CLIENT, "true");
+        System.setProperty(clientX509Util.getSslKeystoreLocationProperty(), testDataPath + "/ssl/testKeyStore.jks");
+        System.setProperty(clientX509Util.getSslKeystorePasswdProperty(), "testpass");
+        System.setProperty(clientX509Util.getSslTruststoreLocationProperty(), testDataPath + "/ssl/testTrustStore.jks");
+        System.setProperty(clientX509Util.getSslTruststorePasswdProperty(), "testpass");
+        //System.setProperty("zookeeper.authProvider.x509", "org.apache.zookeeper.server.auth.SFDCX509AuthenticationProvider");
+        //System.setProperty(clientX509Util.getSslProviderProperty(), DynamicKeyStoreProvider.PKIKS_PROVIDER_NAME);
+    }
+
+    @After
+    public void teardown() throws Exception {
+        System.clearProperty(ServerCnxnFactory.ZOOKEEPER_SERVER_CNXN_FACTORY);
+        System.clearProperty(ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET);
+        System.clearProperty(ZKClientConfig.SECURE_CLIENT);
+        System.clearProperty(clientX509Util.getSslKeystoreLocationProperty());
+        System.clearProperty(clientX509Util.getSslKeystorePasswdProperty());
+        System.clearProperty(clientX509Util.getSslTruststoreLocationProperty());
+        System.clearProperty(clientX509Util.getSslTruststorePasswdProperty());
+        clientX509Util.close();
+    }
+
+    /**
+     * Developers might use standalone mode (which is the default for one server).
+     * This test checks metrics for authz failure in standalone server
+     */
+    @Test
+    public void testSecureStandaloneServerAuthZFailure() throws Exception {
+        System.setProperty("zookeeper.ssl.allowedRoles", "testauthz");
+        Integer secureClientPort = PortAssignment.unique();
+        MainThread mt = new MainThread(QuorumPeerTestBase.MainThread.UNSET_MYID, "", secureClientPort, false);
+        mt.start();
+
+        try {
+            ZooKeeper zk = createZKClnt("127.0.0.1:" + secureClientPort);
+        } catch (Exception e){
+            ServerStats serverStats = mt.getSecureCnxnFactory().getZooKeeperServer().serverStats();
+            assertTrue(serverStats.getAuthFailedCount()>=1);
+        }
+        mt.shutdown();
+
+    }
+
+    private ZooKeeper createZKClnt(String cxnString) throws Exception {
+        ClientBase.CountdownWatcher watcher = new ClientBase.CountdownWatcher();
+        ZooKeeper zk = new ZooKeeper(cxnString, TIMEOUT, watcher);
+        watcher.waitForConnected(CONNECTION_TIMEOUT);
+        return zk;
+    }
+
+    public static class MainThread extends Thread {
```

Comment:
why do you need this class ?

Reply:
second this. We have same class from `QuorumPeerTestBase` - can we reuse that instead of copy paste?

Reply:
FWIW, this test actually has more refrence to `ZooKeeperServerMainTest , i'll use MainThread from that test and add my changes there, and remove the copy and paste code